### PR TITLE
Add Filter and Sort classes, search bar

### PR DESCRIPTION
**Pull request by varnan6**
- **BookFilter.java:** Contains classes returning books sorted by genre and title
- **BookSort.java:** Contains classes sorting books on number of reviews and lexicographical title order
- **Book.java**: Using HashSet to store genres
- **books_data.json**: Smallcased existing genre titles

### CODE REVIEW

File: src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import com.hardvan.springboot.springolearn.model.Book;
 import com.hardvan.springboot.springolearn.service.BookService;
+import com.hardvan.springboot.springolearn.service.BookFilter;
```
Comment (Hardvan) [inline src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java:5]:
```suggestion
import com.hardvan.springboot.springolearn.service.BookFilter;
import com.hardvan.springboot.springolearn.service.BookService;
```

The imports should be in ascending order.

File: src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java [hunk 2]
```diff
@@ -35,4 +39,22 @@ public String deleteAllReviews(@RequestParam String title) {
         bookService.deleteAllReviews(title);
         return "redirect:/";
     }
+
+    @GetMapping("/search")
+    public String searchBooks(@RequestParam String query, @RequestParam String searchType, Model model) {
```
Comment (Hardvan) [inline src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java:44]: You can add function documentation comments here with the help of [javadoc](https://www.oracle.com/in/technical-resources/articles/java/javadoc-tool.html) to make the function's functionality more readable.

File: src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+        List<Book> booksByGenre = new ArrayList<>();
+        for (Book book : books) {
+            if (book.getGenres().contains(searchQuery.toLowerCase())) {
+                booksByGenre.add(book);
+            }
+        }
+
+        return booksByGenre;
+    }
```
Comment (Hardvan) [inline src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java]: Similarly, you can add javadoc function docstrings here as well.

**Review by Hardvan — CHANGES_REQUESTED** [review]
Good usage of the filter and sort functions.

File: src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java (hunk 1 above)
Comment (varnan6) [reply src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java:5]: Okay. Why so?

File: src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java (hunk 3 above)
Comment (varnan6) [reply src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java]: Okay.

File: src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java (hunk 2 above)
Comment (varnan6) [reply src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java:44]: Oh, okay.

File: src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java (hunk 1 above)
Comment (Hardvan) [reply src/main/java/com/hardvan/springboot/springolearn/controller/BookController.java:5]: Just a java convention.

File: src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java (hunk 3 above)
Comment (Hardvan) [reply src/main/java/com/hardvan/springboot/springolearn/service/BookFilter.java]: Once these changes are done, you can commit & push. I will merge the PR then.

**Comment by varnan6** [comment]
The corresponding changes have been made and have been committed.

**Comment by Hardvan** [comment]
Looks good, thanks!

**Pull request merged**